src/FileCache.php

```diff
@@ -135,27 +172,43 @@ protected function setValue($key, $value, $ttl): bool
         $this->gc();
         $cacheFile = $this->getCacheFile($key);
         if ($this->directoryLevel > 0) {
-            @FileHelper::createDirectory(\dirname($cacheFile), $this->dirMode, true);
+            try {
+                @FileHelper::createDirectory(\dirname($cacheFile), $this->dirMode, true);
+            } catch (\yii\exceptions\Exception $error) {
+                return false;
+            }
         }
         // If ownership differs the touch call will fail, so we try to
         // rebuild the file from scratch by deleting it first
         // https://github.com/yiisoft/yii2/pull/16120
         if (is_file($cacheFile) && \function_exists('posix_geteuid') && fileowner($cacheFile) !== posix_geteuid()) {
             @unlink($cacheFile);
         }
-        if (@file_put_contents($cacheFile, $value, LOCK_EX) !== false) {
-            if ($this->fileMode !== null) {
-                @chmod($cacheFile, $this->fileMode);
-            }
-            if ($ttl <= 0) {
-                $ttl = 31536000; // 1 year
-            }
 
-            return @touch($cacheFile, $ttl + time());
+        try {
+            // this implementation is more accurate than file_put_contents
+            // because it does not modify file content, if failed to read
+            if ($fd = fopen($cacheFile, 'cb'))
+            {
+                if (!flock($fd, LOCK_EX)) {
+                    throw new \RuntimeException("Failed to flock '$cacheFile'");
+                } else if(!ftruncate($fd, 0)) {
+                    throw new \RuntimeException("Failed to truncate '$cacheFile");
+                } else if(file_put_contents($cacheFile, $value) !== StringHelper::byteLength($value)) {
```

Comment:
Does it make sense to use `file_put_contents` if we have opened file descriptor?

Reply:
In order not to write following code in user-space:
```
public function fwriteAll($fp, $str)
{
    if (!mb_orig_strlen($str)) return true;
    $bytes_sent = 0;
    $chunk_size = 65536;
    $write = array($fp);
    $read = $except = null;
    while (stream_select($read, $write, $except, null)) {
        $chunk = substr($str, $bytes_sent, $chunk_size);
        $result = fwrite($fp, $chunk);
        if ($result === false || $result === 0) {
            return false;
        }

        $bytes_sent += $result;
        if ($bytes_sent >= strlen($str)) return true;
        $write = array($fp);
    }

    return true;
}
```